pkg/jobmanager/manager.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobmanager
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/adalton/teleport-exercise/pkg/cgroup/v1"
+	"github.com/adalton/teleport-exercise/pkg/config"
+	"github.com/adalton/teleport-exercise/pkg/io"
+)
+
+const (
+	// Superuser is the name of the user who can access any job.
+	Superuser = "administrator"
+)
+
+// JobConstructor is a type that models a function for creating Jobs.
+// This enables us to have a "real" job constructor function as well as
+// constructor functions for mock job implementations that share the same
+// signature.
+type JobConstructor func(
+	owner string,
+	jobName string,
+	controllers []cgroup.Controller,
+	programPath string,
+	arguments ...string,
+) Job
+
+// Manager maintains the set of jobs and enforces the authorization policy
+type Manager struct {
+	mutex               sync.RWMutex
+	jobsByUserByJobId   map[string]map[string]Job // userId->jobId->job
+	jobsByUserByJobName map[string]map[string]Job // userId->jobName->job
+	allJobsByJobId      map[string]Job            // jobId->job
+	controllers         []cgroup.Controller
+	jobConstructor      JobConstructor
+}
+
+// NewManager creates and returns a new standard Manager.
+func NewManager() *Manager {
+	controllers := []cgroup.Controller{
+		cgroup.NewCpuController().SetCpus(config.CgroupDefaultCpuLimit),
+		cgroup.NewMemoryController().SetLimit(config.CgroupDefaultMemoryLimit),
+		cgroup.NewBlockIoController().
+			SetReadBpsDevice(config.CgroupDefaultBlkioReadLimit).
+			SetWriteBpsDevice(config.CgroupDefaultBlkioWriteLimit),
+	}
+
+	return NewManagerDetailed(NewJob, controllers)
+
+}
+
+// NewManagerDetailed returns a new Manger with the given values.
+// The jobConstructor is a function for creating new jobs.  In production
+// this will point to NewJob.  For unit tests, this might point to a
+// constructor function for a mock type.
+// The given controllers is the list of cgroup controllers to manage while
+// running jobs.
+func NewManagerDetailed(jobConstructor JobConstructor, controllers []cgroup.Controller) *Manager {
+	return &Manager{
+		jobsByUserByJobId:   make(map[string]map[string]Job),
+		jobsByUserByJobName: make(map[string]map[string]Job),
+		allJobsByJobId:      make(map[string]Job),
```

Comment:
I agree that having multiple views into the same data is easy to mess up, and what you're suggesting would work.  There are some tradeoffs to consider.  I think having multiple views has some potential to be easier to read/understand because each directly lines up with a usage pattern and uses abstractions that everyone is already familiar with.  If we build some other abstraction to avoid using multiple maps, then the reader would need to first understand that abstraction and how it's providing the necessary information.  It's not a giant leap, but it is an extra step with some additional risk of misunderstanding.  We have unit tests for this component, so we have some ability to build confidence that we've got the multiple view approach correct, and that it stays correct over time.

It would boil down to a question of is the risk of using multiple views and getting it wrong greater than potentially making the implementation harder to understand.  I don't have a strong opinion, but I tend to prefer things that are easier to read/understand.  If you have a strong preference for your suggestion, I'll be happy to implement it.

Reply:
Sounds fair.